pennylane/circuit_drawer/drawable_layers.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains a helper function to sort operations into layers.
+"""
+
+from .utils import default_wire_map
+
+
+def _recursive_find_layer(checking_layer, op_occupied_wires, occupied_wires_per_layer):
+    """Determine correct layer for operation with ``op_occupied_wires``
+
+    Args:
+        checking_layer (int): function determines if operation fits on this layer
+        op_occupied_wires (set(int)): wires covered the drawn operation.  Includes everything
+            between used wires in a multi-wire gate.
+        occupied_wires_per_layer (list[set[int]]): which wires will already have something drawn
+            on them
+
+    Returns:
+        int: layer to place relevant operation in
+    """
+
+    if occupied_wires_per_layer[checking_layer] & op_occupied_wires:
+        # this layer is occupied, use higher one
+        return checking_layer + 1
+    if checking_layer == 0:
+        # reached first layer, so stop
+        return 0
+    # keep pushing the operation back
+    return _recursive_find_layer(checking_layer - 1, op_occupied_wires, occupied_wires_per_layer)
+
+
+def drawable_layers(ops, wire_map=None):
+    """Determine non-overlapping yet dense placement of operations for drawing.
+
+    Args:
+        ops Iterable[~.Operator]: a list of operations
+
+    Keyword Args:
+        wire_map=None (dict): a map from wire label to non-negative integers
+
+    Returns:
+        list[set[~.Operator]] : Each index is a set of operations
+            for the corresponding layer
+    """
+
+    if wire_map is None:
+        wire_map = default_wire_map(ops)
+
+    # initialize
+    max_layer = 0
+    occupied_wires_per_layer = [set()]
+    ops_per_layer = [set()]
+
+    # loop over operations
+    for op in ops:
+        if len(op.wires) == 0:
+            # if no wires, then it acts of all wires
+            # for example, state and sample
+            mapped_wires = set(wire_map.values())
+            op_occupied_wires = mapped_wires
+        else:
+            mapped_wires = {wire_map[wire] for wire in op.wires}
+            op_occupied_wires = set(range(min(mapped_wires), max(mapped_wires) + 1))
```

Comment:
It could be worth making this line more readable by expressing it in multiple lines with descriptive object names.